src/table.ts

```diff
@@ -9,57 +9,106 @@ import {QueryParams} from './query_params';
 import Record from './record';
 import callbackToPromise from './callback_to_promise';
 import Base from './base';
+import {Records} from './records';
+import {FieldSet} from './field_set';
+import {RecordData} from './record_data';
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+type TableError = any;
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+type CreateRecord<TFields> = Pick<RecordData<Partial<TFields>>, 'fields'>;
+type CreateRecords<TFields> = string[] | Partial<TFields>[] | CreateRecord<TFields>[];
+
+type OptionalParameters = {typecast?: boolean};
+
+type RecordCollectionCallback<TFields extends FieldSet> = (
+    error: TableError,
+    records?: Records<TFields>
+) => void;
+type RecordCallback<TFields extends FieldSet> = (
+    error: TableError,
+    record?: Record<TFields>
+) => void;
+type RecordListCallback<TFields extends FieldSet> = (
+    error: TableError,
+    records?: Records<TFields>,
+    offset?: number
+) => void;
+type RecordForEachCallback<TFields extends FieldSet> = (record: Record<TFields>) => void;
+type RecordForEachDoneCallback = (error?: TableError) => void;
+
+interface TableFindRecord<TFields extends FieldSet> {
+    (recordId: string): Promise<Record<TFields>>;
+    (recordId: string, done: RecordCallback<TFields>): void;
+}
 
-type OptionalParameters = {[key: string]: any};
-
-type RecordCallback = (error: any, record?: Record) => void;
-type RecordCollectionCallback = (error: any, records?: Record[]) => void;
-
-interface TableFindRecord {
-    (recordId: string): Promise<Record>;
-    (recordId: string, done: RecordCallback): void;
+interface TableSelectRecord<TFields extends FieldSet> {
+    (params?: QueryParams<TFields>): Query<TFields>;
 }
 
-interface TableCreateRecords {
-    (recordsData: any[], optionalParameters?: OptionalParameters): Promise<Record[]>;
+interface TableCreateRecords<TFields extends FieldSet> {
+    (recordsData: CreateRecords<TFields>, optionalParameters?: OptionalParameters): Promise<
```

Comment:
`@types/airtable` types `recordsData` as `TFields[]`. #210 types `recordsData` as `string[]` as an example. [//airtable.com/api] types `recordsData` as `{fields: TFields}[]`.

Are all there right? That is currently what `CreateRecords` expands to: `string[] | TFields[] | {fields: TFields}[]`.

Reply:
As well should we type with the given `TFields`? Or should we use `Partial<TFields>`? Using `TFields` directly is more flexible for users, they can opt to make the fields optional. Using `Partial<TFields>` users get optional fields but can't force them to be required.